### PR TITLE
Track rotation and flip - fixes for containWithinAspectRatio

### PR DESCRIPTION
**Pull request by meirg**
Track rotation/flip of original loaded image to fix rotation when containWithinAspectRatio is true, and added transformation info to ImageCroppedEvent

### CODE REVIEW

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 1]
```diff
@@ -150,33 +162,93 @@ export class ImageCropperComponent implements OnChanges {
...
-        this.originalBase64 = imageBase64;
-        this.safeImgDataUrl = this.sanitizer.bypassSecurityTrustResourceUrl(imageBase64);
-        this.originalImage = new Image();
-        this.originalImage.onload = () => {
-            this.originalSize.width = this.originalImage.width;
-            this.originalSize.height = this.originalImage.height;
-            this.cd.markForCheck();
-        };
-        this.originalImage.src = imageBase64;
+    private transformOriginalImage(): Promise<void> {
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: Can you split up this function up into a few smaller ones. Some suggestions below to make it easier to follow what's going on:

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 2]
```diff
@@ -150,33 +162,93 @@ export class ImageCropperComponent implements OnChanges {
...
+        if (!this.originalImage || !this.originalImage.complete ) {
+            return Promise.reject(new Error('No Image Loaded'));
+        }
+        return new Promise<void>((resolve, reject) => {
+            const canvas = document.createElement('canvas');
+            const ctx = canvas.getContext('2d');
+            this.transformations.rotation = this.transformations.rotation % 4;
+            if (this.transformations.rotation < 0) {
+                this.transformations.rotation += 4;
+            }
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: checkRotation()

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 3]
```diff
@@ -150,33 +162,93 @@ export class ImageCropperComponent implements OnChanges {
...
+                }
+            } else {
+                if (this.transformations.rotation % 2) {
+                    canvas.height = this.originalSize.width;
+                    canvas.width = this.originalSize.height;
+                } else {
+                    canvas.width = this.originalSize.width;
+                    canvas.height = this.originalSize.height;
+                }
+            }
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: getTransformSize()

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 4]
```diff
@@ -150,33 +162,93 @@ export class ImageCropperComponent implements OnChanges {
...
+                canvas.width / 2,
+                canvas.height / 2
+            );
+            ctx.rotate(Math.PI * (this.transformations.rotation / 2));
+            ctx.drawImage(
+                this.originalImage,
+                -this.originalSize.width / 2,
+                -this.originalSize.height / 2
+            );
+            this.transformedBase64 = canvas.toDataURL();
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: transformImage()

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 5]
```diff
@@ -150,33 +162,93 @@ export class ImageCropperComponent implements OnChanges {
...
+                this.transformedSize.width = this.transformedImage.width;
+                this.transformedSize.height = this.transformedImage.height;
+                resolve();
+                this.cd.markForCheck();
+            };
+            this.transformedImage.onerror = () => {
+                this.transformedImage = null;
+                reject();
+            };
+            this.transformedImage.src = this.transformedBase64;
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: setTransformedImage()

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 6]
```diff
@@ -150,33 +162,93 @@ export class ImageCropperComponent implements OnChanges {
...
+            ctx.drawImage(
+                this.originalImage,
+                -this.originalSize.width / 2,
+                -this.originalSize.height / 2
+            );
+            this.transformedBase64 = canvas.toDataURL();
+            this.safeImgDataUrl = this.sanitizer.bypassSecurityTrustResourceUrl(this.transformedBase64);
+            this.transformedImage = new Image();
+            this.transformedImage.onload = () => {
+                this.transformedSize.width = this.transformedImage.width;
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: In the originalImage you changed width to naturalWidth and height to naturalHeight, it should be done here as well?

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 1 above)
Comment (meirg) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: I split it up, see if it's better now. Thanks!

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 6 above)
Comment (meirg) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: Can't hurt, they should always be the same though. Changed

**Comment by Mawi137** [comment]
Released in `2.0.1`! Thanks for this feature ;)

**Pull request merged**